### DISCUSSION TITLE
Knowledge Review Method

### DISCUSSION THREAD

**Discussion by AndyFrancesco29** [yoshitomo-matsubara/torchdistill#143] [Q&A]
Hi @yoshitomo-matsubara ,

Thanks for your implementation of knowledge review method. Here are some questions regarding your implementation:

1. I think the layers that need to be hooked in teacher network is a little different. In the original code, it sets the preact to be True. So it actually extract the output before the last relu function in layer (which is the input of the last relu function of last block in one layer). I am not sure if this will effect training result. Also I print the matrix but could hardly find a value less than 0. However it should make more sense to use value before relu in teacher, otherwise it will have some 0s when calculating loss.

2. Also I think in the original code there are four ABF modules. The resnet has structure like this "conv1-layer1-layer2-layer3-avgpool". Although it requires to extract features from five stages but there are four modules (the same number of in_channels). I not sure if there is some misunderstanding or use different notations etc., but since you could run with that so I guess that should be OK?

3. Most of parts in my implementation is same as yours but I put those abf values in the black box. In this example it could be done this way since it is just the output from each abfs module. However in SemCKD it requires to compare teacher and student features shape and decide to do operations on both of them. This is the place I got confused.

Anyway it is still a great work! I really appreciate your help.

**Comment by yoshitomo-matsubara** [comment]
Hi @AndyFrancesco29 

> I think the layers that need to be hooked in teacher network is a little different. In the original code, it sets the preact to be True. So it actually extract the output before the last relu function in layer (which is the input of the last relu function of last block in one layer). I am not sure if this will effect training result. Also I print the matrix but could hardly find a value less than 0. However it should make more sense to use value before relu in teacher, otherwise it will have some 0s when calculating loss.

Thank you for pointing out. Since it looks not described so in their paper, I don't know if it's essential to use the input of the last ReLU call in each of layers 1-4. But, I'll follow their implementation just in case.

> Also I think in the original code there are four ABF modules. The resnet has structure like this "conv1-layer1-layer2-layer3-avgpool". Although it requires to extract features from five stages but there are four modules (the same number of in_channels). I not sure if there is some misunderstanding or use different notations etc., but since you could run with that so I guess that should be OK?

I think you referred to their models for CIFAR-100 (thus, "conv1-layer1-layer2-layer3-avgpool"), but note that my sample configuration is for ImageNet dataset as put under `ilsvrc2012/`

> Most of parts in my implementation is same as yours but I put those abf values in the black box. In this example it could be done this way since it is just the output from each abfs module. However in SemCKD it requires to compare teacher and student features shape and decide to do operations on both of them. This is the place I got confused.

Could you point to where (in their code) you found that "However in SemCKD it requires to compare teacher and student features shape and decide to do operations on both of them. This is the place I got confused" ?

**Reply by AndyFrancesco29** [reply]
Hi @yoshitomo-matsubara ,

> I think you referred to their models for CIFAR-100 (thus, "conv1-layer1-layer2-layer3-avgpool"), but note that my sample configuration is for ImageNet dataset as put under ilsvrc2012/

Oops I miss that. Thanks for clarification.

> Could you point to where (in their code) you found that "However in SemCKD it requires to compare teacher and student features shape and decide to do operations on both of them. This is the place I got confused" ?

This is the [part](https://github.com/DefangChen/SemCKD/blob/master/models/util.py) in SelfA.forward(), #feature space alignment# part. In SemCKD it is not like that in review KD, every layer output in student is associated with each layer in teacher. So sometimes teacher has larger size than student. What it has done was:

1. If student is larger than teacher, pool **student** into same shape. Then use a conv block for **pooled student** to get same channels. Use **pooled-conv-ed student** and **origin teacher** to calculate loss;

2. If teacher is larger than student, pool teacher into same shape. Then use a conv block for **original student** to get same channels. Use **pooled teacher** and **conv-ed student** to get loss.
(Hope this is clear)

So in practice if I put that in yaml I have to specific which teacher and which student to hook. If there are four layers in teacher and four layers in student, in total it needs 4x4=16 configurations.

**Reply by yoshitomo-matsubara** [reply]
Hi @AndyFrancesco29 

> So in practice if I put that in yaml I have to specific which teacher and which student to hook. If there are four layers in teacher and four layers in student, in total it needs 4x4=16 configurations.

It looks like the size comparison between student and teacher outputs only cares [the height of patch size](https://github.com/DefangChen/SemCKD/blob/master/models/util.py#L109).
Since image classification often uses fixed input shape, the intermediate feature map shapes are also fixed given the input shape.
So there will be up to 4 module definitions in yaml and each of them can have a list of binary values to indicate whether its H is bigger than that of teacher.

By the way, I started feeling if the importance of SemCKD method is significant. Table 2 in the paper uses weak baselines (even missing CRD, that is partly used in Table 5). For the complexity of the proposed approach, the ImageNet accuracy with a popular teacher-student pair (ResNet-34 ResNet-18) is comparable to or worse than many of my reproduced experiments (summarized [here](https://github.com/yoshitomo-matsubara/torchdistill#top-1-validation-accuracy-for-ilsvrc-2012-imagenet) and [Table 5 in my paper](https://github.com/yoshitomo-matsubara/torchdistill#citation)).
